tests/trades.test.ts

```diff
@@ -0,0 +1,214 @@
+      priceNumerator: 200000n * 10n ** 18n,
+      priceDenominator: 1000n * 10n ** 18n,
+      maxSellAmount: 200000n * 10n ** 18n,
+      minReceiveAmount: 1000n * 10n ** 18n,
+      soldVolume: 100000n * 10n ** 18n,
+      boughtVolume: 500n * 10n ** 18n,
+      trades: [],
+      createEpoch: 0n,
+      cancelEpoch: null,
+      deleteEpoch: null,
+      txHash: `0x${'00'.repeat(32)}`,
+      txLogIndex: 0n,
+    })
+
+    expect(() =>
+      mappings.onTradeReversion(
+        {
+          owner: user,
+          orderId,
+          sellToken: 0,
+          buyToken: 1,
+          executedSellAmount: 100000n * 10n ** 18n,
+          executedBuyAmount: 500n * 10n ** 18n,
+        },
+        {
+          block: { timestamp: 10n * 300n },
+        },
+      ),
+    ).to.throw('aborted')
+  })
+
+  it('errors if there is more than one active trade for an order', async () => {
+    const mappings = await Mappings.load()
+
+    const user = `0x${'01'.repeat(20)}`
+    const orderId = 1337n
+    const orderUid = `${user}-${orderId}`
+    const solutionTx = `0x${'10'.repeat(32)}`
+    mappings.setEntity('Order', orderUid, {
+      id: orderUid,
+      owner: user,
+      orderId,
+      fromBatchId: 0n,
+      fromEpoch: 0n,
+      untilBatchId: 42n,
+      untilEpoch: 42n * 300n,
+      buyToken: '0',
+      sellToken: '1',
+      priceNumerator: 200000n * 10n ** 18n,
+      priceDenominator: 1000n * 10n ** 18n,
+      maxSellAmount: 200000n * 10n ** 18n,
+      minReceiveAmount: 1000n * 10n ** 18n,
+      soldVolume: 100000n * 10n ** 18n,
+      boughtVolume: 500n * 10n ** 18n,
+      trades: [],
+      createEpoch: 0n,
+      cancelEpoch: null,
+      deleteEpoch: null,
+      txHash: `0x${'00'.repeat(32)}`,
+      txLogIndex: 0n,
+    })
+    mappings.setEntity('Trade', `${user}-1`, {
+      id: `${user}-1`,
+      order: orderUid,
+      owner: user,
+      sellVolume: 100000n * 10n ** 18n,
+      buyVolume: 500n * 10n ** 18n,
+      tradeBatchId: 9n,
+      tradeEpoch: 10n * 300n,
+      buyToken: '0',
+      sellToken: '1',
+      createEpoch: 10n * 300n,
+      revertEpoch: null,
+      txHash: solutionTx,
+      txLogIndex: 1n,
+    })
+    mappings.setEntity('Trade', `${user}-2`, {
+      id: `${user}-2`,
+      order: orderUid,
+      owner: user,
+      sellVolume: 100000n * 10n ** 18n,
+      buyVolume: 500n * 10n ** 18n,
+      tradeBatchId: 9n,
+      tradeEpoch: 10n * 300n,
+      buyToken: '0',
+      sellToken: '1',
+      createEpoch: 10n * 300n,
+      revertEpoch: null,
+      txHash: solutionTx,
+      txLogIndex: 2n,
+    })
+
+    expect(() =>
+      mappings.onTradeReversion(
+        {
+          owner: user,
+          orderId,
+          sellToken: 0,
+          buyToken: 1,
+          executedSellAmount: 100000n * 10n ** 18n,
+          executedBuyAmount: 500n * 10n ** 18n,
+        },
+        {
+          block: { timestamp: 10n * 300n },
+        },
+      ),
+    ).to.throw('aborted')
+  })
+})
```

Comment:
Maybe another nice test would be to revert an order that has already been reverted in the same batch? Like, when you have solution, revert, solution, revert.

Reply:
Note that they will be different trade entities (even if they are for the same order) as each solution creates new "trade" entities. This is still a valid test though, as it verifies that the trade entity associated with the `order` is being correctly chosen.